service/controller/v11/resource/pod/delete.go

```diff
@@ -38,80 +41,201 @@ func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
 
-		for _, f := range currentPod.GetFinalizers() {
-			if f == key.DrainingNodesFinalizer {
-				changed = true
-				continue
+			err := r.createNodeConfig(ctx, currentPod)
+			if err != nil {
+				return microerror.Mask(err)
 			}
 
-			newFinalizers = append(newFinalizers, f)
+			resourcecanceledcontext.SetCanceled(ctx)
+			finalizerskeptcontext.SetKept(ctx)
+			r.logger.LogCtx(ctx, "debug", "canceling reconciliation for pod")
+
+		} else if err != nil {
+			return microerror.Mask(err)
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "found node config for the guest cluster")
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "waiting for inspection of the reconciled pod")
 		}
 
-		if changed {
-			podToDelete = currentPod
-			podToDelete.SetFinalizers(newFinalizers)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "inspecting node config for the guest cluster")
+
+		if !nodeConfig.Status.HasFinalCondition() {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has no final state")
+			resourcecanceledcontext.SetCanceled(ctx)
+			finalizerskeptcontext.SetKept(ctx)
+			r.logger.LogCtx(ctx, "debug", "canceling reconciliation for pod")
+
+			return nil
 		}
-	}
 
-	r.logger.LogCtx(ctx, "debug", "looking if the pod has to be updated or to be deleted in the Kubernetes API")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has final state")
 
-	if podToDelete != nil {
-		r.logger.LogCtx(ctx, "debug", "the pod has to be updated or to be deleted in the Kubernetes API")
+		err = r.deleteNodeConfig(ctx, nodeConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
 
-		{
-			r.logger.LogCtx(ctx, "debug", "updating the pod in the Kubernetes API to remove the pod's 'draining-nodes' finalizer")
+	var podToDelete *corev1.Pod
+	{
+		podToDelete = currentPod
 
-			_, err := r.k8sClient.CoreV1().Pods(podToDelete.Namespace).Update(podToDelete)
-			if errors.IsConflict(err) {
-				// The reconciled pod may be updated by other processes or even humans
-				// meanwhile. In case the resource version we currently know does not
-				// match the latest existing one, we give up here and wait for the
-				// delete event to be replayed. Then we try again later until we
-				// succeed.
-				r.logger.LogCtx(ctx, "debug", "cannot update the pod in the Kubernetes API to remove the pod's 'draining-nodes' finalizer because of outdated resource version")
-				return nil
-			} else if err != nil {
-				return microerror.Mask(err)
-			}
+		a := podToDelete.GetAnnotations()
+		a[key.AnnotationPodDrained] = "True"
+		podToDelete.SetAnnotations(a)
+	}
 
-			r.logger.LogCtx(ctx, "debug", "updated the pod in the Kubernetes API to remove the pod's 'node-drainer' finalizer")
+	{
+		r.logger.LogCtx(ctx, "debug", "updating the pod in the Kubernetes API")
+
+		_, err := r.k8sClient.CoreV1().Pods(podToDelete.Namespace).Update(podToDelete)
```

Comment:
Why trying to remove annotation when pod is deleted in next operation?

Reply:
We are not removing an annotation here. We update it. We set the `AnnotationPodDrained` to `True` because the process of draining is successfully done.